modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,489 @@
+        type-info   (hx/type-info expr)
+        db-type     (hx/type-info->db-type type-info)]
+    (if (and ;; AT TIME ZONE is only valid on these Presto types; if applied to something else (ex: `date`), then
+             ;; an error will be thrown by the query analyzer
+             (contains? #{"timestamp" "timestamp with time zone" "time" "time with time zone"} db-type)
+             ;; if one has already been set, don't do so again
+             (not (::in-report-zone? (meta expr)))
+             report-zone)
+      (-> (hx/with-database-type-info (->AtTimeZone expr report-zone) timestamp-with-time-zone-db-type)
+        (vary-meta assoc ::in-report-zone? true))
+      expr)))
+
+;; most date extraction and bucketing functions need to account for report timezone
+
+(defmethod sql.qp/date [:presto-jdbc :default]
+  [_ _ expr]
+  expr)
+
+(defmethod sql.qp/date [:presto-jdbc :minute]
+  [_ _ expr]
+  (hsql/call :date_trunc (hx/literal :minute) (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :minute-of-hour]
+  [_ _ expr]
+  (hsql/call :minute (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :hour]
+  [_ _ expr]
+  (hsql/call :date_trunc (hx/literal :hour) (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :hour-of-day]
+  [_ _ expr]
+  (hsql/call :hour (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :day]
+  [_ _ expr]
+  (hsql/call :date (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :day-of-week]
+  [_ _ expr]
+  (sql.qp/adjust-day-of-week :presto-jdbc (hsql/call :day_of_week (in-report-zone expr))))
+
+(defmethod sql.qp/date [:presto-jdbc :day-of-month]
+  [_ _ expr]
+  (hsql/call :day (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :day-of-year]
+  [_ _ expr]
+  (hsql/call :day_of_year (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :week]
+  [_ _ expr]
+  (sql.qp/adjust-start-of-week :presto-jdbc (partial hsql/call :date_trunc (hx/literal :week)) (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :month]
+  [_ _ expr]
+  (hsql/call :date_trunc (hx/literal :month) (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :month-of-year]
+  [_ _ expr]
+  (hsql/call :month (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :quarter]
+  [_ _ expr]
+  (hsql/call :date_trunc (hx/literal :quarter) (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :quarter-of-year]
+  [_ _ expr]
+  (hsql/call :quarter (in-report-zone expr)))
+
+(defmethod sql.qp/date [:presto-jdbc :year]
+  [_ _ expr]
+  (hsql/call :date_trunc (hx/literal :year) (in-report-zone expr)))
+
+(defmethod sql.qp/unix-timestamp->honeysql [:presto-jdbc :seconds]
+  [_ _ expr]
+  (let [report-zone (qp.timezone/report-timezone-id-if-supported :presto-jdbc)]
+    (hsql/call :from_unixtime expr (hx/literal (or report-zone "UTC")))))
+
+(defmethod sql.qp/unix-timestamp->honeysql [:presto-jdbc :milliseconds]
```

Comment:
I suppose this is fine as well, but other implementations are fine with just dividing by 1000 and doing a seconds conversion. Seems either is fine.

Reply:
Yeah, I figure that since Presto does support millisecond precision, might as well use it even though it's slightly complicated.